experiments/controller.go

```diff
@@ -231,21 +256,19 @@ func (ec *ExperimentController) syncHandler(key string) error {
 		return err
 	}
 
-	exCtx := experimentContext{
-		log:                    logCtx,
-		ex:                     experiment,
-		templateRSs:            templateRSs,
-		kubeclientset:          ec.kubeclientset,
-		argoProjClientset:      ec.argoProjClientset,
-		replicaSetLister:       ec.replicaSetLister,
-		recorder:               ec.recorder,
-		enqueueExperimentAfter: ec.enqueueExperimentAfter,
-	}
+	exCtx := newExperimentContext(
+		experiment,
+		templateRSs,
+		ec.kubeclientset,
+		ec.argoProjClientset,
+		ec.replicaSetLister,
+		ec.analysisTemplateLister,
+		ec.analysisRunLister,
+		ec.recorder,
+		ec.enqueueExperimentAfter,
+	)
 
-	newStatus, err := exCtx.reconcile()
-	if err != nil {
-		return err
-	}
+	newStatus := exCtx.reconcile()
```

Comment:
This is a fair concern. I will say that fundamentally, we do have to handle things little differently from long lived resources (Rollout) vs. resources with a lifecycle (Experiment). If this were Rollout, I would agree that everything should be fault tolerant, and we should retry by default.

But since this is Experiments, we will need to be more intelligent on when something is retry-able or not, or else we could infinitely retry on things that are not retryable and not fail the experiment as necessary.

Generally speaking, for lifecycle resources I prefer to fail fast, early and often, and then improve the retry capability over time, whereas fault tolerant resources I would start on the other side of spectrum and retry by default.

Note that the way we can handle retries without having to surface errors to the top, is to simply not fail the experiment. This will eventually get retried on the next reconciliation.

For this PR, I would like to continue the fail fast route, but think of ways to improve this in the future.

Reply:
Thinking about this more, I think agree with you about not returning an error. What I now realize, is without returning an error, we won't be able to take advantage of the rate limiting workqueue. Since this PR is huge enough, I've filed an issue to improve this:

https://github.com/argoproj/argo-rollouts/issues/227